### PR TITLE
Fix #62, Fix #314: Integrate DuckDuckGo search engine callout and popup & minor UI issue

### LINKED ISSUE #62: DuckDuckGo Integration Migration
Migrate existing DDG in private mode screens and functionality.

- [ ] DDG popup screen when coming to Private Browsing for the first time
- [ ] DDG label on private browsing 
    - Only visible if DDG is _not_ the default search engine for private mode
    - Tapping on it DDG label re-opens the pop up
- [ ] Popup options must both work (Yes and No)

| Label | Popup |
| --- | --- |
|  |  |

### LINKED ISSUE #314: Tab Tray's toolbar buttons not aligned on devices with a bottom safe area
The buttons should be aligned. The private mode button & done button are not respecting the bottom safe area.

simulator screen shot - iphone xs max - 2018-10-04 at 16 15 22

### PR DESCRIPTION
**Pull request by kylehickinson**
One thing to note, we do not focus the url bar when we create a new private tab at the moment. If we decide to change this, we will need to add the necessary checks to assure focus is not done when a popup is going to be shown.

Also: `PopupView` and `AlertPopupView` are almost 100% ported aside from the need to replace all instances of `getApp()` with a new local convenience getter `applicationWindow`

## Pull Request Checklist

- [x] My patch has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!`
- [ ] I have updated the *Unit Tests* to cover new or changed functionality
- [ ] I have updated the *UI Tests* to cover new or changed functionality
- [x] I have made sure that localizable strings use `NSLocalizableString()`

## Screenshots

_None included_

## Notes for testing this patch

_None included_

### CODE REVIEW

**Review by iccub — CHANGES_REQUESTED** [review]
Doesn't work well in landscape mode, this also affects 1.6.

zrzut ekranu 2018-10-9 o 16 38 35

I ran it and the pr works okay, will do a code review later(I know most of it is copy pasted from 1.6)

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift [hunk 1]
```diff
@@ -50,11 +52,34 @@ class FavoritesViewController: UIViewController {
...
+    }()
+    
+    @objc private func showDDGCallout() {
+        delegate?.didTapDuckDuckGoCallout()
+    }
     
     // MARK: - Init/lifecycle
-    init() {
+    
+    let profile: Profile
```
Comment (jhreis) [inline Client/Frontend/Browser/HomePanel/FavoritesViewController.swift]: private?

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift [hunk 2]
```diff
@@ -50,11 +52,34 @@ class FavoritesViewController: UIViewController {
         $0.autoresizingMask = [.flexibleWidth]
     }
     
-    /// Called after user taps on ddg popup to set it as a default search enginge in private browsing mode.
-    var ddgPrivateSearchCompletionBlock: (() -> Void)?
+    fileprivate var ddgLogo = UIImageView(image: #imageLiteral(resourceName: "duckduckgo"))
+    
+    fileprivate lazy var ddgLabel: UILabel = {
```
Comment (jhreis) [inline Client/Frontend/Browser/HomePanel/FavoritesViewController.swift]: Use `then`?

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift [hunk 3]
```diff
@@ -50,11 +52,34 @@ class FavoritesViewController: UIViewController {
...
+        let label = UILabel()
+        label.numberOfLines = 0
+        label.textColor = BraveUX.GreyD
+        label.font = UIFont.systemFont(ofSize: 14, weight: UIFont.Weight.regular)
+        label.text = Strings.DDG_promotion
+        return label
+    }()
+    
+    fileprivate lazy var ddgButton: UIControl = {
```
Comment (jhreis) [inline Client/Frontend/Browser/HomePanel/FavoritesViewController.swift]: `then`?

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift [hunk 4]
```diff
@@ -104,8 +130,15 @@ class FavoritesViewController: UIViewController {
...
         
         collection.addSubview(braveShieldStatsView)
+        collection.addSubview(ddgButton)
+        
+        ddgButton.addSubview(ddgLogo)
+        ddgButton.addSubview(ddgLabel)
         
         makeConstraints()
+        
+        collection.addObserver(self, forKeyPath: #keyPath(UICollectionView.contentSize), options: [.new, .initial], context: nil)
```
Comment (jhreis) [inline Client/Frontend/Browser/HomePanel/FavoritesViewController.swift]: Any reason not to use the newer block-based observing?

Comment (kylehickinson) [reply Client/Frontend/Browser/HomePanel/FavoritesViewController.swift]: Na, just entirely out of habit lol. Updated!

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift [hunk 5]
```diff
@@ -50,11 +52,30 @@ class FavoritesViewController: UIViewController {
...
+    private let ddgLogo = UIImageView(image: #imageLiteral(resourceName: "duckduckgo"))
+    
+    private let ddgLabel = UILabel().then {
+        $0.numberOfLines = 0
+        $0.textColor = BraveUX.GreyD
+        $0.font = UIFont.systemFont(ofSize: 14, weight: UIFont.Weight.regular)
+        $0.text = Strings.DDG_promotion
+    }
+    
+    private lazy var ddgButton = UIControl().then {
```
Comment (jhreis) [inline Client/Frontend/Browser/HomePanel/FavoritesViewController.swift:64]: remove `lazy`

Comment (kylehickinson) [reply Client/Frontend/Browser/HomePanel/FavoritesViewController.swift:64]: Can't, it uses `self` in its initialization (adding a target)

Comment (jhreis) [reply Client/Frontend/Browser/HomePanel/FavoritesViewController.swift:64]: Fine, be smart.

**Review by jhreis — APPROVED** [review]
Sick!

**Comment by jhreis** [comment]
@iccub landscape is an existing issue, so no point on blocking on this right now IMO. Create a new ticket for it and we'll deal with it after 1.7.0.

**Comment by iccub** [comment]
Fine by me.

**Review by iccub — APPROVED** [review]

**Pull request merged**